### PR TITLE
JavaScript: Restructure implementation of `DataFlow::SourceNode`.

### PR DESCRIPTION
**Pull request by xiemaisi**
It now uses a facade pattern similar to `InvokeNode`: the range of the class is defined by an abstract class `DataFlow::SourceNode::Range`, while the actual behaviour is defined by the (no longer abstract) `SourceNode` class itself.

Clients that want to add new source nodes need to extend `DataFlow::SourceNode::Range`, those that want to refine the behaviour of existing source nodes should extend `DataFlow::SourceNode` itself.

While this is technically a breaking API change, I think separating the two aspects in this way is cleaner and makes it easier to use, and improves [performance](https://git.semmle.com/gist/max/032a0aa577e8f6d9b5a488981cc4f6a3) (internal link) as well.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll [hunk 1]
```diff
@@ -246,8 +246,9 @@ class GlobalVarRefNode extends DataFlow::ValueNode, DataFlow::DefaultSourceNode
  */
 DataFlow::SourceNode globalObjectRef() {
   // top-level `this`
-  exists (ThisNode globalThis | result = globalThis |
-    not exists(globalThis.getBinder())
+  exists (StmtContainer sc |
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: This change seems unrelated to this PR.

File: javascript/ql/src/semmle/javascript/dataflow/Sources.qll [hunk 2]
```diff
@@ -9,15 +9,17 @@
...
- * A source node for local data flow, that is, a node for which local
- * data flow cannot provide any information about its inputs.
+ * A source node for local data flow, that is, a node from which local data flow is tracked.
  *
- * By default, functions, object and array expressions and JSX nodes
- * are considered sources, as well as expressions that have non-local
- * flow (such as calls and property accesses). Additional sources
- * can be modelled by extending this class with additional subclasses.
+ * Examples include function parameters, imports and property accesses; see
+ * `DataFlow::SourceNode::DefaultRange` for details. You can introduce new kinds of
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/dataflow/Sources.qll]: Minor:
Consistent docstring style: the use of "you" is very rare in ql/qll files. Should the sentence be something like the following instead?

> Define new subclasses of `DataFlow::SourceNode::Range` to introduce new kinds of source nodes.

File: javascript/ql/src/semmle/javascript/dataflow/Sources.qll [hunk 3]
```diff
@@ -187,44 +189,63 @@ abstract class SourceNode extends DataFlow::Node {
...
- *   - object expressions
- *   - array expressions
- *   - JSX literals.
- */
-class DefaultSourceNode extends SourceNode {
+module SourceNode {
+  /**
+   * A data flow node that should be considered a source node.
+   *
+   * Subclass this class to introduce new kinds of source nodes. If you want to refine
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/dataflow/Sources.qll]: Ditto regarding "you".

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll (hunk 1 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: In a sense, yes. It is, however, necessary to prevent a negative recursion.

Comment (esben-semmle) [reply javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: :+1:

**Comment by xiemaisi** [comment]
>  the introduction of yet another naming scheme (...) is a bit unfortunate

I agree. Which of the other ones would you like me to switch to?

**Comment by asger-semmle** [comment]
I like this change a lot. I also like the `::Range` better than the other conventions so far, so maybe the change others to use that.

Long ago I started a write-up on a canonical naming scheme [here](https://docs.google.com/document/d/1Nyhbo-Ont-QMMIWAoU9KA3nvcIEoWeInxEXSwd0DddU/edit?usp=sharing) but I gave up on rolling it out after the miserable performance I saw in the string concatenation library.

**Comment by esben-semmle** [comment]
I also prefer the `::Range` scheme, let's use that from now on, and eventually migrate the other schemes.

File: javascript/ql/src/semmle/javascript/dataflow/Sources.qll (hunk 2 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/dataflow/Sources.qll]: I modelled it loosely after the docstring for `InvokeNodeDef`, which also has "you".

**Comment by xiemaisi** [comment]
I've rebased to resolve conflicts with the autoformatting PR.

Two further minor API-related questions are perhaps worth discussing:

  - Should we introduce an empty `abstract class CustomSourceNode extends SourceNode::Range` and tell people to extend that instead of `SourceNode::Range` itself? The name might be slightly more intuitive, but of course it's functionally much the same as just extending `SourceNode::Range` itself.

  - Should we introduce a field `SourceNode::Range impl` in `SourceNode` like we have in `InvokeNode`? It wouldn't currently be useful for anything since `SourceNode::Range` doesn't implement any member predicates so we don't need `impl` outside the characteristic predicate, but it would be a step towards a more unified design pattern.
